File: cli-go/cmd/doctor.go
package cmd

import (
	"errors"
	"fmt"
	"io"
	"time"

	"github.com/jedib0t/go-pretty/v6/table"
	"github.com/spf13/cobra"

	"github.com/imprezahost/impreza-devkit/cli-go/internal/client"
	"github.com/imprezahost/impreza-devkit/cli-go/internal/config"
	"github.com/imprezahost/impreza-devkit/cli-go/internal/output"
)

// `impreza doctor` runs five sequenced health checks. Exit 0 only if
// all pass. Mirrors the Python CLI's doctor — same check names, same
// labels, same exit semantics.

const (
	doctorOK   = "[OK]"
	doctorFail = "[FAIL]"
	doctorWarn = "[WARN]"
	doctorSkip = "[SKIP]"
)

type doctorCheck struct {
	Name    string `json:"name"`
	Status  string `json:"status"` // OK | FAIL | WARN | SKIP (no brackets)
	Summary string `json:"summary"`
	Detail  string `json:"detail,omitempty"`
	OK      bool   `json:"ok"`
}

var doctorCmd = &cobra.Command{
	Use:   "doctor",
	Short: "Five-check health verification. Exits 0 only if all checks pass.",
	Long: `Sequenced health check:

  1. active-context: a context is configured + selected (no network).
  2. api-reachable:  GET /account/api-keys/self returns 200 OK.
  3. key-status:     the API key is active (not paused / revoked).
  4. ip-whitelist:   the request_ip the server saw is on the key whitelist.
  5. account-profile: GET /account returns the profile + balance.

Each check renders as [OK] / [FAIL] / [WARN] / [SKIP]. Exit code:

  0 — every check passed
  1 — any check failed

Use this as a smoke test after install ("does my context work?"),
in CI ("are credentials still valid?"), or as a first-line support
artefact ("paste the impreza doctor output").`,
	Args: cobra.NoArgs,
	RunE: func(cmd *cobra.Command, args []string) error {
		out := cmd.OutOrStdout()
		checks := runDoctorChecks(cmd)

		f, err := resolveFormat()
		if err != nil {
			return err
		}
		if f != output.FormatTable {
			return renderJSONOrYAML(out, checks, f)
		}

		writeDoctorReport(out, checks)

		// Exit non-zero if any check failed.
		failed := 0
		for _, c := range checks {
			if !c.OK {
				failed++
			}
		}
		if failed > 0 {
			return fmt.Errorf("doctor: %d of %d checks failed", failed, len(checks))
		}
		return nil
	},
}

func runDoctorChecks(cmd *cobra.Command) []doctorCheck {
	out := []doctorCheck{}

	// ── 1. active-context (no network) ────────────────────────────
	cfg, err := config.Load()
	if err != nil {
		out = append(out, doctorCheck{
			Name:    "active-context",
			Status:  "FAIL",
			Summary: "No config file found",
			Detail:  fmt.Sprintf("%v\n\nRun `impreza context create <name> --key imp_... --secret ...` to set up your first context.", err),
			OK:      false,
		})
		// Subsequent checks need a context; mark them SKIP.
		for _, n := range []string{"api-reachable", "key-status", "ip-whitelist", "account-profile"} {
			out = append(out, doctorCheck{Name: n, Status: "SKIP", Summary: "no active context"})
		}
		return out
	}
	name, ctx, err := cfg.Active(globalContext)
	if err != nil {
		out = append(out, doctorCheck{
			Name:    "active-context",
			Status:  "FAIL",
			Summary: "No default context set and no --context override",
			Detail:  fmt.Sprintf("%v\n\nRun `impreza context use <name>` or pass --context <name>.", err),
			OK:      false,
		})
		for _, n := range []string{"api-reachable", "key-status", "ip-whitelist", "account-profile"} {
			out = append(out, doctorCheck{Name: n, Status: "SKIP", Summary: "no active context"})
		}
		return out
	}
	out = append(out, doctorCheck{
		Name:    "active-context",
		Status:  "OK",
		Summary: fmt.Sprintf("Active context: %s", name),
		OK:      true,
	})

	// ── 2. api-reachable + 3. key-status + 4. ip-whitelist
	//      (all three are derived from the same GET .../api-keys/self call).
	c, err := client.New(ctx)
	if err != nil {
		out = append(out, doctorCheck{
			Name: "api-reachable", Status: "FAIL",
			Summary: "Couldn't build HTTP client", Detail: err.Error(), OK: false,
		})
		return appendSkipChecks(out, "key-status", "ip-whitelist", "account-profile")
	}

	start := time.Now()
	id, err := c.ApiKeySelf(cmd.Context())
	elapsed := time.Since(start).Round(time.Millisecond)

	if err != nil {
		var ae *client.APIError
		if errors.As(err, &ae) {
			switch ae.Status {
			case 401:
				out = append(out, doctorCheck{
					Name: "api-reachable", Status: "FAIL",
					Summary: "Authentication failed (HTTP 401)",
					Detail:  fmt.Sprintf("%s. The API key or secret is invalid. Rotate via your Impreza Account, update the local context with `impreza context create / use`.", ae.Message),
					OK:      false,
				})
			case 403:
				out = append(out, doctorCheck{
					Name: "api-reachable", Status: "FAIL",
					Summary: "Forbidden (HTTP 403)",
					Detail:  fmt.Sprintf("%s. Add the calling IP to the key's whitelist via your Impreza Account, or use a different key whose whitelist already covers this IP.", ae.Message),
					OK:      false,
				})
			default:
				out = append(out, doctorCheck{
					Name: "api-reachable", Status: "FAIL",
					Summary: fmt.Sprintf("HTTP %d", ae.Status),
					Detail:  ae.Error(),
					OK:      false,
				})
			}
		} else {
			out = append(out, doctorCheck{
				Name: "api-reachable", Status: "FAIL",
				Summary: "Network error", Detail: err.Error(), OK: false,
			})
		}
		return appendSkipChecks(out, "key-status", "ip-whitelist", "account-profile")
	}

	out = append(out, doctorCheck{
		Name:    "api-reachable",
		Status:  "OK",
		Summary: fmt.Sprintf("GET /account/api-keys/self OK (%s)", elapsed),
		Detail:  fmt.Sprintf("key prefix=%q, label=%q", id.Prefix, id.Label),
		OK:      true,
	})

	// ── 3. key-status
	keyOK := id.Status == "active"
	keyStatus := "OK"
	if !keyOK {
		keyStatus = "FAIL"
	}
	out = append(out, doctorCheck{
		Name:    "key-status",
		Status:  keyStatus,
		Summary: fmt.Sprintf("status=%q", id.Status),
		Detail: func() string {
			if keyOK {
				return ""
			}
			return "Pending, paused, or revoked keys will start returning 401 unpredictably. Rotate via your Impreza Account."
		}(),
		OK: keyOK,
	})

	// ── 4. ip-whitelist
	if id.RequestIP == "" {
		out = append(out, doctorCheck{
			Name:    "ip-whitelist",
			Status:  "WARN",
			Summary: "Server did not report request_ip",
			Detail:  "The /api-keys/self endpoint should echo the IP it saw the request from. Missing field is unusual but the request authenticated; inspect via your Impreza Account to be sure.",
			OK:      true, // warn but don't fail
		})
	} else {
		matched := false
		matchLabel := ""
		for _, e := range id.IPWhitelist {
			if e.IPAddress == id.RequestIP {
				matched = true
				matchLabel = e.Label
				if matchLabel == "" {
					matchLabel = "(no label)"
				}
				break
			}
		}
		if matched {
			out = append(out, doctorCheck{
				Name:    "ip-whitelist",
				Status:  "OK",
				Summary: fmt.Sprintf("request_ip %s matches entry (%s)", id.RequestIP, matchLabel),
				OK:      true,
			})
		} else {
			labels := []string{}
			for _, e := range id.IPWhitelist {
				labels = append(labels, e.IPAddress)
			}
			out = append(out, doctorCheck{
				Name:    "ip-whitelist",
				Status:  "FAIL",
				Summary: fmt.Sprintf("request_ip %s not in whitelist (%d entr%s)", id.RequestIP, len(id.IPWhitelist), pluralY(len(id.IPWhitelist))),
				Detail:  fmt.Sprintf("Whitelist: %v. Add the calling IP via your Impreza Account, or switch to a context whose key already allows this IP.", labels),
				OK:      false,
			})
		}
	}

	// ── 5. account-profile
	info, err := c.AccountInfo(cmd.Context())
	if err != nil {
		out = append(out, doctorCheck{
			Name: "account-profile", Status: "FAIL",
			Summary: "Couldn't fetch /account", Detail: err.Error(), OK: false,
		})
	} else {
		out = append(out, doctorCheck{
			Name:    "account-profile",
			Status:  "OK",
			Summary: fmt.Sprintf("%s %s <%s>, balance %.2f %s", info.FirstName, info.LastName, info.Email, info.Balance, info.Currency),
			Detail:  fmt.Sprintf("registered %s", info.RegisteredAt),
			OK:      true,
		})
	}

	return out
}

// pluralY returns "y" or "ies" for the singular/plural of "entry".
func pluralY(n int) string {
	if n == 1 {
		return "y"
	}
	return "ies"
}

// appendSkipChecks appends SKIP rows for the listed check names. Used
// when an earlier check fails and the rest can't usefully run.
func appendSkipChecks(checks []doctorCheck, names ...string) []doctorCheck {
	for _, n := range names {
		checks = append(checks, doctorCheck{
			Name: n, Status: "SKIP", Summary: "earlier check failed",
		})
	}
	return checks
}

// writeDoctorReport renders the table output with a top + bottom rule.
func writeDoctorReport(w io.Writer, checks []doctorCheck) {
	fmt.Fprintln(w, "impreza doctor")
	fmt.Fprintln(w, "----------------------------------------")
	t := table.NewWriter()
	t.SetOutputMirror(w)
	t.SetStyle(table.StyleDefault)
	style := t.Style()
	style.Options.DrawBorder = false
	style.Options.SeparateRows = false
	style.Options.SeparateColumns = false
	style.Options.SeparateHeader = false
	for _, c := range checks {
		label := doctorLabel(c.Status)
		row := []any{label, c.Name + ": " + c.Summary}
		t.AppendRow(row)
		if c.Detail != "" {
			t.AppendRow([]any{"", "        " + c.Detail})
		}
	}
	t.Render()
	fmt.Fprintln(w, "----------------------------------------")
	passed, failed := 0, 0
	for _, c := range checks {
		if c.Status == "SKIP" {
			continue
		}
		if c.OK {
			passed++
		} else {
			failed++
		}
	}
	if failed == 0 {
		fmt.Fprintf(w, "All checks passed. %d/%d.\n", passed, passed)
	} else {
		fmt.Fprintf(w, "FAILED: %d of %d checks failed.\n", failed, passed+failed)
	}
}

func doctorLabel(status string) string {
	switch status {
	case "OK":
		return doctorOK
	case "FAIL":
		return doctorFail
	case "WARN":
		return doctorWarn
	case "SKIP":
		return doctorSkip
	default:
		return "[" + status + "]"
	}
}

func init() {
	rootCmd.AddCommand(doctorCmd)
}

File: cli-python/impreza_cli/commands/doctor.py
"""``impreza doctor`` health-check command — Phase 4.1.

A single-command first-line support tool. Runs a sequence of
checks against the resolved context and reports the state of
each, so a user troubleshooting "I just installed and nothing
works" gets a copy-pasteable diagnostic instead of digging
through eight different verbs.

Checks (in order, stopping early on hard failures):

1. **Config file resolved** — does ``~/.config/impreza/config.toml``
   exist and parse? (Implicit via ``make_client_or_exit``; failures
   land as the standard config-error stderr.)
2. **Active context** — which context resolved? Show its label
   and the API key prefix.
3. **API reachable** — round-trip ``GET /account/api-keys/self``.
   Network errors and auth errors both surface here.
4. **Key status** — the returned ``KeyIdentity.status`` field.
5. **IP whitelist** — does the server-observed ``request_ip`` match
   one of the whitelist entries? Mismatch is the single most
   common "everything is 403" cause.
6. **Account profile + balance** — round-trip ``GET /account``.
   Sanity check that the key has actual scopes beyond
   ``/account/api-keys/self``.

Each check renders as ``[OK] / [FAIL] / [WARN]`` (ASCII-only —
no Unicode glyphs, per the Phase 1.6 cp1252 lesson). The exit
code is 0 if every check passed, 1 otherwise. Output mode
``--output json`` emits a structured array suitable for piping
into monitoring scripts.
"""

from __future__ import annotations

import json
import sys
import time
from typing import Any

import typer
from impreza.exceptions import (
    ApiError,
    AuthError,
    IpNotWhitelisted,
    NetworkError,
    PermissionDenied,
)

from ..output import OutputFormat, error, success, warning
from ..sdk import make_client_or_exit
from ..state import from_typer_context, resolve_output

app = typer.Typer(
    name="doctor",
    help="Run a health check against the active context.",
    invoke_without_command=True,
    no_args_is_help=False,
)


# ── data model ──────────────────────────────────────────────────────


class _CheckResult:
    """One row in the doctor report. Mutable so callers can build it
    in-place across the check body.

    ``ok`` is True/False/None — None means "skipped" (typically
    because an earlier check failed and this one depends on it).
    """

    __slots__ = ("name", "ok", "summary", "detail")

    def __init__(self, name: str) -> None:
        self.name: str = name
        self.ok: bool | None = False
        self.summary: str = ""
        self.detail: str = ""

    def passed(self, summary: str, detail: str = "") -> None:
        self.ok = True
        self.summary = summary
        self.detail = detail

    def failed(self, summary: str, detail: str = "") -> None:
        self.ok = False
        self.summary = summary
        self.detail = detail

    def warned(self, summary: str, detail: str = "") -> None:
        # WARN renders distinctly but still counts as "passed" for
        # the overall exit code. Used for cosmetic mismatches that
        # don't actually break functionality.
        self.ok = True
        self.summary = "WARN: " + summary
        self.detail = detail

    def skipped(self, summary: str) -> None:
        self.ok = None
        self.summary = summary
        self.detail = ""

    def as_dict(self) -> dict[str, Any]:
        return {
            "name": self.name,
            "ok": self.ok,
            "summary": self.summary,
            "detail": self.detail,
        }


# ── individual checks ───────────────────────────────────────────────


def _check_context(state: Any) -> _CheckResult:
    """Report which context resolved. Doesn't actually verify the
    API key — that's the next check's job. The ``make_client_or_exit``
    call in the parent already failed-fast on missing-config errors,
    so reaching this check at all means there's a usable context."""
    r = _CheckResult("active-context")
    override = state.context_override
    if override:
        r.passed(f"Context override: {override!r}")
    else:
        r.passed("Default context")
    return r


def _check_api_reachable(client: Any, key_holder: dict[str, Any]) -> _CheckResult:
    """Round-trip GET /account/api-keys/self. Captures elapsed time
    so the report shows latency, and stashes the result in
    ``key_holder`` so later checks reuse it without a second HTTP."""
    r = _CheckResult("api-reachable")
    t0 = time.monotonic()
    try:
        key = client.account.api_key_self()
    except NetworkError as exc:
        r.failed(
            "Could not reach api.imprezahost.com",
            f"Network error: {exc}. Check connectivity / DNS / proxy. "
            "If using Tor, confirm the SOCKS proxy is up.",
        )
        return r
    except AuthError as exc:
        r.failed(
            "Authentication failed (HTTP 401)",
            f"{exc.message}. The API key or secret is invalid. "
            "Rotate via Impreza Account, update the local context "
            "with `impreza context create / use`.",
        )
        return r
    except IpNotWhitelisted as exc:
        r.failed(
            "IP not whitelisted (HTTP 403)",
            f"{exc.message}. Add the calling IP to the API key's "
            "whitelist via Impreza Account, or use a different "
            "key whose whitelist already covers this IP.",
        )
        return r
    except PermissionDenied as exc:
        r.failed(
            "Permission denied (HTTP 403)",
            f"{exc.message}.",
        )
        return r
    except ApiError as exc:
        r.failed(
            f"API error: {exc.message}",
            f"code={exc.code or '?'}, status={exc.status_code or '?'}",
        )
        return r
    elapsed_ms = (time.monotonic() - t0) * 1000.0
    key_holder["key"] = key
    r.passed(
        f"GET /account/api-keys/self OK ({elapsed_ms:.0f}ms)",
        f"key prefix={key.prefix!r}, label={key.label or '(unnamed)'!r}",
    )
    return r


def _check_key_status(key_holder: dict[str, Any]) -> _CheckResult:
    """Inspect the KeyIdentity.status field returned by api_key_self.
    Active keys pass; anything else (paused, revoked, etc.) flags as
    a hard fail — even though the call succeeded, the key may stop
    working at any time."""
    r = _CheckResult("key-status")
    key = key_holder.get("key")
    if key is None:
        r.skipped("api-reachable check failed; cannot inspect key")
        return r
    status = (key.status or "").lower()
    if status == "active":
        r.passed(f"status={status!r}")
    else:
        r.failed(
            f"status={key.status!r} (not active)",
            "Pending, paused, or revoked keys will start returning "
            "401 unpredictably. Rotate via Impreza Account.",
        )
    return r


def _check_ip_whitelist(key_holder: dict[str, Any]) -> _CheckResult:
    """Compare ``request_ip`` (what the server saw the call coming
    from) to ``ip_whitelist`` (the entries the server would accept)."""
    r = _CheckResult("ip-whitelist")
    key = key_holder.get("key")
    if key is None:
        r.skipped("api-reachable check failed; cannot inspect whitelist")
        return r

    request_ip = key.request_ip or ""
    entries = list(key.ip_whitelist or [])
    if not request_ip:
        r.warned(
            "server did not echo request_ip in this response",
            "Whitelist check is unavailable; the call already "
            "succeeded so the IP must be allowed, but the report "
            "cannot prove it.",
        )
        return r
    if not entries:
        r.warned(
            f"request_ip {request_ip} reached the API but the key "
            "has no whitelist entries",
            "Either the key has whitelist enforcement disabled "
            "(unusual) or the server is letting it through anyway. "
            "Inspect via Impreza Account to be sure.",
        )
        return r

    match = next((e for e in entries if e.ip_address == request_ip), None)
    if match is None:
        labels = ", ".join(
            f"{e.ip_address!r}{f' ({e.label!r})' if e.label else ''}"
            for e in entries
        )
        r.failed(
            f"request_ip {request_ip} not in whitelist "
            f"({len(entries)} entr{'y' if len(entries) == 1 else 'ies'})",
            f"Whitelist: {labels}. Add the calling IP via your Impreza "
            "Account, or switch to a context whose key already "
            "allows this IP.",
        )
        return r

    label = f" ({match.label!r})" if match.label else ""
    r.passed(f"request_ip {request_ip} matches entry{label}")
    return r


def _check_account_profile(client: Any) -> _CheckResult:
    """Round-trip GET /account. The api_key_self endpoint sometimes
    bypasses scope checks; calling /account exercises a regular
    read-scope so the doctor catches scope-limited keys early."""
    r = _CheckResult("account-profile")
    try:
        acc = client.account.get()
    except ApiError as exc:
        r.failed(
            f"GET /account failed: {exc.message}",
            f"code={exc.code or '?'}, status={exc.status_code or '?'}. "
            "If api-reachable passed but this didn't, the key may "
            "lack the basic read scope. Contact support.",
        )
        return r
    name = f"{acc.first_name} {acc.last_name}".strip()
    if acc.company:
        name = f"{name} ({acc.company})"
    r.passed(
        f"{name} <{acc.email}>, balance {acc.balance:.2f} {acc.currency}",
        f"registered {acc.registered_at}",
    )
    return r


# ── renderers ───────────────────────────────────────────────────────


_LABEL_COLOR = {
    True: typer.colors.GREEN,
    False: typer.colors.RED,
    None: typer.colors.YELLOW,
}
_LABEL_TEXT = {True: "[OK]  ", False: "[FAIL]", None: "[SKIP]"}


def _is_warn(result: _CheckResult) -> bool:
    return result.ok is True and result.summary.startswith("WARN:")


def _print_text_report(results: list[_CheckResult]) -> None:
    """Write the human-friendly report to stdout. Failures and warns
    get their detail indented under the summary line so a copy-paste
    of the whole report preserves the full context."""
    sys.stdout.write("\n")
    sys.stdout.write("impreza doctor\n")
    sys.stdout.write("-" * 40 + "\n")
    for r in results:
        if _is_warn(r):
            label = "[WARN]"
            color = typer.colors.YELLOW
        else:
            label = _LABEL_TEXT[r.ok]
            color = _LABEL_COLOR[r.ok]
        typer.secho(label, fg=color, bold=True, nl=False)
        sys.stdout.write(f"  {r.name}: {r.summary}\n")
        if r.detail:
            for line in r.detail.splitlines():
                sys.stdout.write(f"        {line}\n")
    sys.stdout.write("-" * 40 + "\n")


def _print_summary(results: list[_CheckResult]) -> int:
    """Bottom-line summary. Returns the exit code (0 if all pass).

    Uses the :mod:`commands.output` palette helpers (4.3) so the
    summary line matches the colour conventions every other CLI
    command will adopt going forward: success = green, warning =
    yellow, error = red.
    """
    total = len(results)
    failed = [r for r in results if r.ok is False]
    skipped = [r for r in results if r.ok is None]
    warned = [r for r in results if _is_warn(r)]
    passed = total - len(failed) - len(skipped)
    if failed:
        error(f"{len(failed)} of {total} checks failed.")
        return 1
    if warned or skipped:
        msg_parts = [f"{passed} of {total} checks passed"]
        if warned:
            msg_parts.append(f"{len(warned)} warning(s)")
        if skipped:
            msg_parts.append(f"{len(skipped)} skipped")
        warning(", ".join(msg_parts) + ".")
        return 0
    success(f"All checks passed. {passed}/{total}.")
    return 0


# ── entry point ─────────────────────────────────────────────────────


@app.callback(invoke_without_command=True)
def doctor(
    typer_ctx: typer.Context,
    output: OutputFormat | None = typer.Option(
        None,
        "--output",
        "-o",
        help=(
            "Output format. 'json' emits the check array structured "
            "for monitoring scripts (each entry has name / ok / "
            "summary / detail). Default: human-friendly text report."
        ),
        case_sensitive=False,
    ),
) -> None:
    """Run a health check against the active context.

    Reports each step (config / context / API reachable / key
    status / IP whitelist / account) and exits 0 only if every
    check passed. ``--output json`` produces a structured report
    suitable for piping into monitoring scripts.

    Common failure modes the doctor catches:

    * Config file missing or empty → first-time setup hint.
    * API key invalid → rotate via your Impreza Account.
    * IP not whitelisted → add the IP or switch context.
    * Key paused / revoked → rotate.
    * Account-scope mismatch → contact support.
    """
    # If a subcommand was invoked, do nothing — Typer routes there.
    if typer_ctx.invoked_subcommand is not None:
        return

    state = from_typer_context(typer_ctx)
    fmt = resolve_output(state, output)

    results: list[_CheckResult] = []

    # Check 1: config + context (failures here exit before we can
    # build a Client; the parent helper renders its own error).
    results.append(_check_context(state))

    # Build the client. make_client_or_exit handles config errors
    # itself with a friendly stderr + Exit(1), so reaching this point
    # means we have a valid client.
    with make_client_or_exit(state) as client:
        key_holder: dict[str, Any] = {}
        results.append(_check_api_reachable(client, key_holder))
        results.append(_check_key_status(key_holder))
        results.append(_check_ip_whitelist(key_holder))
        # Account profile only makes sense if API was reachable.
        if key_holder.get("key") is not None:
            results.append(_check_account_profile(client))
        else:
            r = _CheckResult("account-profile")
            r.skipped("api-reachable failed; cannot test other endpoints")
            results.append(r)

    if fmt is OutputFormat.JSON:
        sys.stdout.write(
            json.dumps([r.as_dict() for r in results], indent=2, ensure_ascii=False)
        )
        sys.stdout.write("\n")
        # JSON consumers want exit-code-as-signal too.
        if any(r.ok is False for r in results):
            raise typer.Exit(code=1)
        return

    _print_text_report(results)
    exit_code = _print_summary(results)
    if exit_code != 0:
        # Don't suppress the report — raise after printing so the
        # user sees the full diagnostic AND gets the non-zero exit.
        # typer.Exit raises a SystemExit(exit_code) which click /
        # CliRunner translates to result.exit_code.
        raise typer.Exit(code=exit_code)
    # Some terminals expect a trailing newline before the prompt.
    # The summary printer already emitted one via secho.

    # Note: we don't use `error()` for failed checks because the
    # report itself contains the [FAIL] markers in red — adding a
    # stderr "Error:" line on top would just duplicate the signal.
